src/java/org/jivesoftware/openfire/nio/NIOConnection.java

```diff
@@ -261,7 +255,15 @@ public boolean isSecure() {
 
     public void deliver(Packet packet) throws UnauthorizedException {
         if (isClosed()) {
-            backupDeliverer.deliver(packet);
+        	// OF-857: Do not allow the backup deliverer to recurse
+        	if (backupDeliverer == null) {
+        		Log.error("Failed to deliver packet: " + packet.toXML());
+        		throw new IllegalStateException("Connection closed");
+        	}
+        	// attempt to deliver via backup only once
+        	PacketDeliverer backup = backupDeliverer;
```

Comment:
Note: I added this "belt and suspenders" code to prevent recursion because the ComponentConnectionHandler still uses the standard PacketDeliverer as a backup.